### PR TITLE
Adds a graphviz plot to visualize the interactions between library commits

### PR DESCRIPTION
**Pull request by cormensratio**
Adds the `b_multi_lib_interaction_graphviz` plot to visualize the interactions between library commits.

### CODE REVIEW

File: varats/varats/data/reports/blame_report.py [hunk 1]
```diff
@@ -516,6 +516,38 @@ def generate_degree_tuples(
...
+    for func_entry in report.function_entries:
+        for interaction in func_entry.interactions:
+            amount = interaction.amount
+            base_commit_repo_pair = CommitRepoPair(
+                interaction.base_commit.commit_hash,
+                interaction.base_commit.repository_name
+            )
+
+            # Add each interaction `amount` times
+            for _ in range(amount):
```
Comment (cormensratio) [inline varats/varats/data/reports/blame_report.py]: I just realised that i could maybe remove this duplication of interactions in the database by writing each amount value to the interaction as extra column. I will try to implement it asap.

Comment (cormensratio) [reply varats/varats/data/reports/blame_report.py]: Done

File: varats/varats/data/databases/blame_library_interactions_database.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+                for inter_pair in inter_pair_amount_dict:
+                    current_rev_data_dict = build_dataframe_row(
+                        base_hash=base_pair.commit_hash,
+                        base_library=base_pair.repository_name,
+                        inter_hash=inter_pair.commit_hash,
+                        inter_library=inter_pair.repository_name,
+                        amount=inter_pair_amount_dict[inter_pair]
+                    )
+                    result_data_dicts.append(current_rev_data_dict)
```
Comment (vulder) [inline varats/varats/data/databases/blame_library_interactions_database.py]: You could just directly append `current_rev_data_dict` without a local var.

File: varats/varats/data/reports/blame_report.py [hunk 3]
```diff
@@ -531,6 +531,44 @@ def generate_degree_tuples(
     return list(degree_dict.items())
 
 
+InteractingCommitRepoPairToAmountMapping = tp.Dict[CommitRepoPair, int]
+
+
+def gen_base_to_inter_commit_repo_pair_mapping(
+    report: tp.Union[BlameReport, BlameReportDiff]
+) -> tp.Dict[CommitRepoPair, InteractingCommitRepoPairToAmountMapping]:
+    """
```
Comment (vulder) [inline varats/varats/data/reports/blame_report.py:540]: Please add description to the doc string.

File: varats/varats/plots/blame_interaction_degree.py [hunk 4]
```diff
@@ -622,6 +652,171 @@ def _build_sankey_figure(
...
+            weight = row['amount']
+            weight_string = str(weight)
+            if not edge_weight_threshold or (
+                weight >= edge_weight_threshold.value
+            ):
+                label = weight_string
+
+        graph.edge(
+            f'{base_hash}_{base_lib}', f'{inter_hash}_{inter_lib}', label=label
+        )
```
Comment (vulder) [inline varats/varats/plots/blame_interaction_degree.py]: Wasn't the idea to leave out edges that don't go over the threshold? This currently looks like only setting a label is dependent on the threshold but not if the edge is addded.

File: varats/varats/plots/blame_interaction_degree.py [hunk 5]
```diff
@@ -622,6 +652,171 @@ def _build_sankey_figure(
...
+        save_path: tp.Optional[Path] = None,
+        filetype: str = 'pdf'
+    ) -> tp.Optional[Digraph]:
+        plot_cfg = {
+            'fig_title': 'MISSING figure title',
+        }
+        if extra_plot_cfg is not None:
+            plot_cfg.update(extra_plot_cfg)
+
+        style.use(self.style)
```
Comment (vulder) [inline varats/varats/plots/blame_interaction_degree.py]: Is this needed for graphviz? Looks like a copy past error from the other matplotlib functions.

File: varats/varats/plots/blame_interaction_degree.py [hunk 6]
```diff
@@ -1120,6 +1321,58 @@ def calc_missing_revisions(self, boundary_gradient: float) -> tp.Set[str]:
...
+        self, path: tp.Optional[Path] = None, filetype: str = 'png'
+    ) -> None:
+        try:
+            self.plot(False)
+        except PlotDataEmpty:
+            LOG.warning(f"No data for project {self.plot_kwargs['project']}.")
+            return
+
+    def calc_missing_revisions(self, boundary_gradient: float) -> tp.Set[str]:
+        pass
```
Comment (vulder) [inline varats/varats/plots/blame_interaction_degree.py]: Should probably raise an NotImplementerError

**Review by vulder — APPROVED** [review]
Overall looks fine to me :)

File: varats/varats/plots/blame_interaction_degree.py (hunk 4 above)
Comment (cormensratio) [reply varats/varats/plots/blame_interaction_degree.py]: Right, I misread your message as we talked about this. It's fixed now.

File: varats/varats/plots/blame_interaction_degree.py (hunk 5 above)
Comment (cormensratio) [reply varats/varats/plots/blame_interaction_degree.py]: It's a copy paste error. I removed the plot_cfg as well, as it is not used in the graphviz plot.

File: varats/varats/data/reports/blame_report.py [hunk 7]
```diff
@@ -531,6 +531,47 @@ def generate_degree_tuples(
...
+    Maps the base CommitRepoPair of an blame interaction to each distinct
```
Comment (boehmseb) [inline varats/varats/data/reports/blame_report.py]:
```suggestion
    Maps the base CommitRepoPair of a blame interaction to each distinct
```

File: varats/varats/data/databases/blame_library_interactions_database.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
+"""Module for the base BlameLibraryInteractionsDatabase class."""
```
Comment (boehmseb) [inline varats/varats/data/databases/blame_library_interactions_database.py]:
```suggestion
"""Module for the BlameLibraryInteractionsDatabase class."""
```

**Review by vulder — APPROVED** [review]

**Review by boehmseb — APPROVED** [review]
Just two typos.

**Pull request merged**